File: go/standard.go
package shape

// Standard Schema V1 interoperability (https://standardschema.dev/).
//
// The Standard Schema spec is a JavaScript/TypeScript convention: a schema
// exposes a "~standard" property so any tool can consume it. Go can't expose
// that JS property, so Shape provides the equivalent surface via Standard():
// a version, a vendor, and a non-throwing Validate that returns either the
// produced value or a list of issues (never both) — mirroring the TS
// `Shape(...)['~standard']` object.

// StandardSchema is the Standard Schema V1-style interface for a compiled shape.
type StandardSchema struct {
	Version  int                            // always 1
	Vendor   string                         // always "shape"
	Validate func(input any) StandardResult // non-throwing validation
}

// StandardResult is the outcome of StandardSchema.Validate: on success Value is
// set and Issues is empty; on failure Issues is populated. Mirrors the TS
// `~standard.validate()` result.
type StandardResult struct {
	Value  any             // produced value (defaults injected) when Issues is empty
	Issues []StandardIssue // validation problems; empty on success
}

// StandardIssue mirrors a Standard Schema V1 issue: a human-readable message and
// the path to the offending value (array indices as ints, object keys as
// strings, matching FieldError.PathArr).
type StandardIssue struct {
	Message string
	Path    []any
}

// Standard returns the Standard Schema V1-style interface for this schema. The
// returned Validate never panics; it reports failures as issues.
func (s *Schema) Standard() StandardSchema {
	return StandardSchema{
		Version: 1,
		Vendor:  "shape",
		Validate: func(input any) StandardResult {
			out, err := s.Validate(input)
			if verr, ok := err.(*ValidationError); ok && verr.hasAny() {
				issues := make([]StandardIssue, len(verr.Issues))
				for i, e := range verr.Issues {
					issues[i] = StandardIssue{Message: e.Text, Path: e.PathArr}
				}
				return StandardResult{Issues: issues}
			}
			return StandardResult{Value: out}
		},
	}
}

File: rs/src/standard.rs
//! Standard Schema V1 interoperability (https://standardschema.dev/): a
//! version, a vendor, and a validate that never fails, reporting either the
//! produced value or a list of issues.

use crate::context::PathPart;
use crate::value::Value;
use crate::Schema;

/// The Standard Schema surface of a compiled shape.
#[derive(Clone, Debug)]
pub struct StandardSchema<'s> {
    /// Always 1.
    pub version: u32,
    /// Always "shape".
    pub vendor: &'static str,
    schema: &'s Schema,
}

/// A Standard Schema issue: a message and the path to the offending value.
#[derive(Clone, Debug, PartialEq)]
pub struct StandardIssue {
    pub message: String,
    pub path: Vec<PathPart>,
}

/// The outcome of a Standard Schema validation: the produced value, or the
/// issues, never both.
#[derive(Clone, Debug, PartialEq)]
pub struct StandardResult {
    pub value: Option<Value>,
    pub issues: Vec<StandardIssue>,
}

impl<'s> StandardSchema<'s> {
    /// Validate without failing: the issues are the result.
    pub fn validate(&self, input: Value) -> StandardResult {
        match self.schema.validate(input) {
            Ok(value) => StandardResult {
                value: Some(value),
                issues: Vec::new(),
            },
            Err(e) => StandardResult {
                value: None,
                issues: e
                    .issues
                    .into_iter()
                    .map(|i| StandardIssue {
                        message: i.text,
                        path: i.path_arr,
                    })
                    .collect(),
            },
        }
    }
}

impl Schema {
    /// The Standard Schema V1 surface of this schema.
    pub fn standard(&self) -> StandardSchema<'_> {
        StandardSchema {
            version: 1,
            vendor: "shape",
            schema: self,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::node::Token;
    use crate::spec::obj;
    use crate::Spec;

    #[test]
    fn reports_issues_or_a_value() {
        let s = Schema::new(obj([("a", Spec::from(Token::Number))]));
        let std = s.standard();
        assert_eq!((std.version, std.vendor), (1, "shape"));
        let ok = std.validate(Value::from(serde_json::json!({"a": 1})));
        assert!(ok.issues.is_empty() && ok.value.is_some());
        let bad = std.validate(Value::from(serde_json::json!({"a": "x"})));
        assert!(bad.value.is_none());
        assert_eq!(bad.issues[0].path, vec![PathPart::Key("a".into())]);
        assert!(bad.issues[0].message.contains("not of type number"));
    }
}
